File: java/src/main/java/it/giannibombelli/wsc2026/common/domain/primitive/Money.java
package it.giannibombelli.wsc2026.common.domain.primitive;

import it.giannibombelli.wsc2026.common.utils.Require;

import java.math.BigDecimal;
import java.math.RoundingMode;

public record Money(BigDecimal value) {
    public Money {
        Require.requireArgument(value, "value");
        try {
            value = value.setScale(2, RoundingMode.UNNECESSARY);
        } catch (ArithmeticException e) {
            throw new IllegalArgumentException("value must have at most 2 decimal places");
        }
        if (value.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("value must not be negative");
        }
    }

    public boolean isPositive() {
        return value.compareTo(BigDecimal.ZERO) > 0;
    }

    public boolean isZero() {
        return value.compareTo(BigDecimal.ZERO) == 0;
    }

    public boolean isLessThan(Money other) {
        return this.value.compareTo(other.value) < 0;
    }

    public boolean isGreaterThan(Money other) {
        return this.value.compareTo(other.value) > 0;
    }

    public Money plus(Money other) {
        return new Money(this.value.add(other.value));
    }

    public Money minus(Money other) {
        return new Money(this.value.subtract(other.value));
    }

    public static Money zero() {
        return new Money(BigDecimal.ZERO);
    }
}

File: typescript/src/common/domain/primitive/money.ts
import { requireArgument } from '@/common/utils/requireArgument.js';
function hasMoreThanTwoDecimals(value: number): boolean {
  const str = value.toString();
  const decimalPointIndex = str.indexOf('.');
  if (decimalPointIndex === -1) {
    return false;
  }
  return str.length - decimalPointIndex - 1 > 2;
}

export class Money {
  readonly value: number;

  constructor(value: number) {
    requireArgument(value, 'value');

    if (value < 0) {
      throw new Error('value must not be negative');
    }

    if (hasMoreThanTwoDecimals(value)) {
      throw new Error('value must have at most 2 decimal places');
    }

    this.value = value;
  }

  isPositive(): boolean {
    return this.value > 0;
  }

  isZero(): boolean {
    return this.value === 0;
  }

  isLessThan(other: Money): boolean {
    return this.value < other.value;
  }

  isGreaterThan(other: Money): boolean {
    return this.value > other.value;
  }

  equals(other: Money): boolean {
    return this.value === other.value;
  }

  plus(other: Money): Money {
    return new Money(this.value + other.value);
  }

  minus(other: Money): Money {
    return new Money(this.value - other.value);
  }

  static zero(): Money {
    return new Money(0);
  }
}
